### PR TITLE
Use kni-installer to deploy masters

### PR DESCRIPTION
**Pull request by stbenjam**
Combined with https://github.com/openshift-metalkube/kni-installer/pull/25, will deploy the masters using kni-installer instead of the shell script.

The hardware information is provided to the installer via a flattened version of ironic_nodes.json. I'm not super happy about how that looks, but the limitation of terraform (see https://github.com/openshift-metalkube/kni-installer/pull/25/files#diff-ff320d88a7ec5e5d76645cd4b4a1c28e) in handling structured data doesn't give us a ton of options.

### CODE REVIEW

**Comment by derekhiggins** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/341/

**Comment by derekhiggins** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/346/

**Comment by hardys** [comment]
Just testing this, note for anyone else trying that the kni-installer PR must be rebased to pick up the latest mdns changes or the masters won't come up.

I guess we'll have to land this and the kni-installer patch at the same time to ensure CI continues working - @derekhiggins do we have any way to pull in dependent changes to ensure all works OK when we do that?

File: 03_ocp_repo_sync.sh [hunk 1]
```diff
@@ -30,7 +30,7 @@ function sync_go_repo_and_patch {
     popd
 }
 
-sync_go_repo_and_patch github.com/openshift-metalkube/kni-installer https://github.com/openshift-metalkube/kni-installer.git
+sync_go_repo_and_patch github.com/openshift-metalkube/kni-installer https://github.com/openshift-metalkube/kni-installer.git https://github.com/openshift-metalkube/kni-installer/pull/25.patch
```
Comment (stbenjam) [inline 03_ocp_repo_sync.sh]: This applies the kni-installer PR temporarily, to get a good CI run hopefully

**Comment by stbenjam** [comment]
Rebased both PR's.  There's a hack in this PR in 03 to patch the kni-installer (see comment above), so we can at least see if CI runs successfully.  

I did have an issue on Friday -- I'm not sure kni-installer will successfully complete.  I removed the early exit (https://github.com/openshift-metalkube/kni-installer/pull/25/files#diff-9ffecdd580b2cb3a1e0e5c2fc92faccfL107), and some operators don't come up like ingress the installer actually fails (very late in the progress - the masters themselves are deployed fine).

**Comment by derekhiggins** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/369/

**Comment by hardys** [comment]
> I did have an issue on Friday -- I'm not sure kni-installer will successfully complete. I removed the early exit (https://github.com/openshift-metalkube/kni-installer/pull/25/files#diff-9ffecdd580b2cb3a1e0e5c2fc92faccfL107), and because some operators don't come up like ingress the installer actually fails (very late in the progress - the masters themselves are deployed fine).

Yeah I did wonder about this - I've seen the same locally so I wonder if we should just move the early exit instead of removing it?

We know the ingress stuff is TODO atm, I think @yboaron is working on it, so we'll need a way to make progress without that for now.

**Comment by yboaron** [comment]
@hardys, Yep, I'm working on adding support for the ingress.
I need that the ingress controller will be deployed successfully. 

My plan is to have the router pods on masters-nodes (host networking),
add support for Ingress VIP (VRRP between masters), and wildcard external DNS
entry point to the ingress VIP.

**Comment by derekhiggins** [comment]
Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/379/

**Comment by hardys** [comment]
Sorry needs a rebase but otherwise lgtm, tested locally and seems to work fine, so I'd say lets rebase then merge the kni-installer patch, and this one right after?

**Comment by derekhiggins** [comment]
> I guess we'll have to land this and the kni-installer patch at the same time to ensure CI continues working - @derekhiggins do we have any way to pull in dependent changes to ensure all works OK when we do that?

Afraid not, sorry.

**Comment by stbenjam** [comment]
Rebased - I removed the hack for the patch as well, so this will fail CI.  I do want to test locally on my box real quick end-to-end, make sure I didnt botch the rebase

**Comment by derekhiggins** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/411/

**Comment by stbenjam** [comment]
It installed:

image

**Comment by derekhiggins** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/412/

**Comment by derekhiggins** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/413/

**Comment by stbenjam** [comment]
Better - should be good to go 

image

**Comment by stbenjam** [comment]
Rebased, and it does deploy the masters.  I'm stuck at the moment with the ingress controller not coming up but it sounds like https://github.com/openshift-metalkube/dev-scripts/pull/222 fixes that and not related to this PR

image

**Comment by derekhiggins** [comment]
Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/487/

**Comment by stbenjam** [comment]
^ That was the CI from this morning, so encouraging news.  Doesn't cover the latest rebase (although nothing significantly changed)

**Comment by hardys** [comment]
Ok tested locally and merged #222, so I think if CI passes here we're good to land the kni-installer patch, remove it from this PR then merge this

**Comment by derekhiggins** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/489/

**Comment by stbenjam** [comment]
Ok, that failure was without #222, so I've rebased it.

**Comment by derekhiggins** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/492/

**Comment by derekhiggins** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/498/

**Comment by hardys** [comment]
I retriggered CI, hopefully https://github.com/openshift-metalkube/kni-installer/pull/33 landing will help

**Comment by markmc** [comment]
This all looks pretty good and I'm keen to see it merged ASAP, leaving any improvements to further iterations. I'm going to play with this for a bit myself now.

The main thing that needs more thought and work as a future iteration is the way we passed the nodes data in install-config.yaml - @stbenjam calls this out in the dev doc too

**Comment by derekhiggins** [comment]
Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/503/

**Comment by stbenjam** [comment]
@hardys Passed! Just removed the patch, these should be good to merge

**Pull request merged**